File: app/src/main/java/com/djavid/bitcoinrate/model/ApiInterface.java
package com.djavid.bitcoinrate.model;

import com.djavid.bitcoinrate.model.dto.blockchain.BlockchainModel;
import com.djavid.bitcoinrate.model.dto.coinmarketcap.CoinMarketCapTicker;
import com.djavid.bitcoinrate.model.dto.cryptonator.CryptonatorTicker;
import com.djavid.bitcoinrate.model.dto.cryptonator.CurrenciesModel;
import com.djavid.bitcoinrate.model.dto.cryptowatch.HistoryDataModel;
import com.djavid.bitcoinrate.model.dto.cryptowatch.PairsResult;
import com.djavid.bitcoinrate.model.dto.heroku.ResponseId;
import com.djavid.bitcoinrate.model.dto.heroku.ResponseTickers;
import com.djavid.bitcoinrate.model.dto.heroku.Subscribe;
import com.djavid.bitcoinrate.model.dto.heroku.Ticker;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;


public interface ApiInterface {

    String HEROKU_URL = "https://bitcoinrate-server.herokuapp.com";


    //blockchain

    @GET("https://blockchain.info/ru/charts/market-price")
    Single<BlockchainModel> getChartValues(@Query("timespan") String timespan,
                                         @Query("sampled") boolean sampled,
                                         @Query("format") String format);


    //cryptonator

    @GET("https://api.cryptonator.com/api/full/{curr1}-{curr2}")
    Single<CryptonatorTicker> getRate(@Path("curr1") String curr1, @Path("curr2") String curr2);

    @GET("https://www.cryptonator.com/api/currencies")
    Single<CurrenciesModel> getCurrencies();


    //coinmarketcap

    @GET("https://api.coinmarketcap.com/v1/ticker/{crypto_id}")
    Single<List<CoinMarketCapTicker>> getRateCMC(@Path("crypto_id") String crypto_id, @Query("convert") String country_id);


    //cryptowatch

    @GET("https://api.cryptowat.ch/markets/{market}/{curr}/ohlc")
    Single<HistoryDataModel> getHistory(@Path("market") String market, @Path("curr") String pair,
                                              @Query("periods") int periods,
                                              @Query("after") long after);

    @GET("https://api.cryptowat.ch/pairs/{pair}")
    Single<PairsResult> getCryptowatchMarkets(@Path("pair") String pair);


    //heroku server

    @GET(HEROKU_URL + "/registerToken")
    Single<ResponseId> registerToken(@Query("token") String token, @Query("id") long id);



    @POST(HEROKU_URL + "/subscribe")
    Single<ResponseId> sendSubscribe(@Body Subscribe subscribe);

    @GET(HEROKU_URL + "/getSubscribes")
    Single<List<Subscribe>> getSubscribesByTokenId(@Header("Token") String token,
                                                   @Query("token_id") long token_id);

    @GET(HEROKU_URL + "/deleteSubscribe")
    Completable deleteSubscribe(@Header("Token") String token, @Query("id") long id);



    @POST(HEROKU_URL + "/addTicker")
    Single<ResponseId> sendTicker(@Body Ticker ticker);

    @GET(HEROKU_URL + "/getTickers")
    Single<ResponseTickers> getTickersByTokenId(@Header("Token") String token,
                                                @Query("token_id") long token_id);

    @GET(HEROKU_URL + "/getTicker")
    Single<Ticker> getTickerByTokenIdAndTickerId(@Header("Token") String token,
                                                 @Query("token_id") long token_id,
                                                 @Query("ticker_id") long ticker_id);

    @GET(HEROKU_URL + "/deleteTicker")
    Completable deleteTicker(@Header("Token") String token, @Query("id") long id);

}

File: app/src/main/kotlin/com/djavid/bitcoinrate/network/ApiInterface.kt
package com.djavid.bitcoinrate.network

import com.djavid.bitcoinrate.model.blockchain.BlockchainModel
import com.djavid.bitcoinrate.model.coinmarketcap.CoinMarketCapTicker
import com.djavid.bitcoinrate.model.cryptocompare.HistoData
import com.djavid.bitcoinrate.model.cryptonator.CryptonatorTicker
import com.djavid.bitcoinrate.model.cryptonator.CurrenciesModel
import com.djavid.bitcoinrate.model.cryptowatch.HistoryDataModel
import com.djavid.bitcoinrate.model.cryptowatch.PairsResult
import com.djavid.bitcoinrate.model.heroku.ResponseId
import com.djavid.bitcoinrate.model.heroku.ResponseTickers
import com.djavid.bitcoinrate.model.heroku.Subscribe
import com.djavid.bitcoinrate.model.heroku.Ticker

import io.reactivex.Completable
import io.reactivex.Single
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query


interface ApiInterface {

    @get:GET("https://www.cryptonator.com/api/currencies")
    val currencies: Single<CurrenciesModel>


    //blockchain

    @GET("https://blockchain.info/ru/charts/market-price")
    fun getChartValues(@Query("timespan") timespan: String,
                       @Query("sampled") sampled: Boolean,
                       @Query("format") format: String): Single<BlockchainModel>


    //cryptonator

    @GET("https://api.cryptonator.com/api/full/{curr1}-{curr2}")
    fun getRate(@Path("curr1") curr1: String, @Path("curr2") curr2: String): Single<CryptonatorTicker>


    //coinmarketcap

    @GET("https://api.coinmarketcap.com/v1/ticker/{crypto_id}")
    fun getRateCMC(@Path("crypto_id") crypto_id: String, @Query("convert") country_id: String): Single<List<CoinMarketCapTicker>>


    //cryptowatch

    @GET("https://api.cryptowat.ch/markets/{market}/{curr}/ohlc")
    fun getHistory(@Path("market") market: String, @Path("curr") pair: String,
                   @Query("periods") periods: Int,
                   @Query("after") after: Long): Single<HistoryDataModel>

    @GET("https://api.cryptowat.ch/pairs/{pair}")
    fun getCryptowatchMarkets(@Path("pair") pair: String): Single<PairsResult>


    //cryptocompare

    @GET("https://min-api.cryptocompare.com/data/histominute")
    fun getHistoMinute(@Query("fsym") from_symbol: String, @Query("tsym") to_symbol: String,
                       @Query("limit") limit: Int, @Query("aggregate") periods: Int,
                       @Query("e") exchange: String): Single<HistoData>

    @GET("https://min-api.cryptocompare.com/data/histohour")
    fun getHistoHour(@Query("fsym") from_symbol: String, @Query("tsym") to_symbol: String,
                     @Query("limit") limit: Int, @Query("aggregate") periods: Int,
                     @Query("e") exchange: String): Single<HistoData>

    @GET("https://min-api.cryptocompare.com/data/histoday")
    fun getHistoDay(@Query("fsym") from_symbol: String, @Query("tsym") to_symbol: String,
                    @Query("limit") limit: Int, @Query("aggregate") periods: Int,
                    @Query("e") exchange: String): Single<HistoData>


    //heroku server

    @GET("$HEROKU_URL/registerToken")
    fun registerToken(@Query("token") token: String, @Query("id") id: Long): Single<ResponseId>


    @POST("$HEROKU_URL/subscribe")
    fun sendSubscribe(@Body subscribe: Subscribe): Single<ResponseId>

    @GET("$HEROKU_URL/getSubscribes")
    fun getSubscribesByTokenId(@Header("Token") token: String,
                               @Query("token_id") token_id: Long): Single<List<Subscribe>>

    @GET("$HEROKU_URL/deleteSubscribe")
    fun deleteSubscribe(@Header("Token") token: String, @Query("id") id: Long): Completable


    @POST("$HEROKU_URL/addTicker")
    fun sendTicker(@Body ticker: Ticker): Single<ResponseId>

    @GET("$HEROKU_URL/getTickers")
    fun getTickersByTokenId(@Header("Token") token: String,
                            @Query("token_id") token_id: Long): Single<ResponseTickers>

    @GET("$HEROKU_URL/getTicker")
    fun getTickerByTokenIdAndTickerId(@Header("Token") token: String,
                                      @Query("token_id") token_id: Long,
                                      @Query("ticker_id") ticker_id: Long): Single<Ticker>

    @GET("$HEROKU_URL/deleteTicker")
    fun deleteTicker(@Header("Token") token: String, @Query("id") id: Long): Completable

    companion object {

        val HEROKU_URL = "https://bitcoinrate-server.herokuapp.com"
    }

}
